ql/instruments/nonstandardswaption.hpp

```diff
@@ -32,62 +32,109 @@
 #include <ql/termstructures/volatility/swaption/swaptionvolstructure.hpp>
 #include <ql/models/calibrationhelper.hpp>
 #include <ql/utilities/disposable.hpp>
+#include <ql/experimental/fx/proxyinstrument.hpp>
 
 namespace QuantLib {
 
-    //! nonstandard swaption class
-    /*! \ingroup instruments
-    */
-
-    class NonstandardSwaption : public Option {
-      public:
-        class arguments;
-        class engine;
-        NonstandardSwaption(const Swaption &fromSwaption);
-        NonstandardSwaption(const boost::shared_ptr<NonstandardSwap> &swap,
-                            const boost::shared_ptr<Exercise> &exercise,
-                            Settlement::Type delivery = Settlement::Physical);
-
-        //! \name Instrument interface
-        //@{
-        bool isExpired() const;
-        void setupArguments(PricingEngine::arguments *) const;
-        //@}
-        //! \name Inspectors
-        //@{
-        VanillaSwap::Type type() const { return swap_->type(); }
-
-        const boost::shared_ptr<NonstandardSwap> &underlyingSwap() const {
-            return swap_;
-        }
-        //@}
-        Disposable<std::vector<boost::shared_ptr<CalibrationHelper> > >
-        calibrationBasket(
-            boost::shared_ptr<SwapIndex> standardSwapBase,
-            boost::shared_ptr<SwaptionVolatilityStructure> swaptionVolatility,
-            const BasketGeneratingEngine::CalibrationBasketType basketType =
-                BasketGeneratingEngine::MaturityStrikeByDeltaGamma) const;
-
-      private:
-        // arguments
-        boost::shared_ptr<NonstandardSwap> swap_;
-        Settlement::Type settlementType_;
-    };
+//! nonstandard swaption class
+/*! \ingroup instruments
+*/
 
-    //! %Arguments for nonstandard swaption calculation
-    class NonstandardSwaption::arguments : public NonstandardSwap::arguments,
-                                           public Option::arguments {
-      public:
-        arguments() {}
-        boost::shared_ptr<NonstandardSwap> swap;
-        Settlement::Type settlementType;
-        void validate() const;
+class NonstandardSwaption : public Option, public ProxyInstrument {
+  public:
+    class arguments;
+    class results;
+    class engine;
+
+    //! proxy description
+    struct Proxy : ProxyDescription {
```

Comment:
I confess I don't have much in mind 😄 
I was trying to wrap my head around the instrument in the core library defining its `results` structure to contain the proxy used by the experimental engine, but maybe I should just stop worrying and learn to love the experimental folder.

I guess the proxy engine would require that the proxies be of a specific type, right? So maybe the proxy engine might define it and the generating engines would create instances. I'm still a bit ambivalent about the proxy being in the instrument results, but it might go in the additional results. What do you think?